modules/currency-exchange/src/lib.rs

```diff
@@ -0,0 +1,403 @@
+// Copyright 2019-2020 Parity Technologies (UK) Ltd.
+// This file is part of Parity Bridges Common.
+
+// Parity Bridges Common is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity Bridges Common is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity Bridges Common.  If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use codec::Encode;
+use frame_support::{decl_error, decl_module, decl_storage, ensure, fail, Parameter};
+use primitives::exchange::{
+	Airdrop, CurrencyConverter, Error as ExchangeError, MaybeLockFundsTransaction, RecipientsMap,
+};
+use sp_runtime::DispatchResult;
+use sp_std::vec::Vec;
+
+/// Peer blockhain interface.
+pub trait Blockchain {
+	/// Block hash type.
+	type BlockHash: Parameter;
+	/// Transaction type.
+	type Transaction: Parameter;
+	/// Transaction inclusion proof type.
+	type TransactionInclusionProof: Parameter;
+
+	/// Verify that transaction is a part of given block.
+	fn verify_transaction_inclusion_proof(
+		transaction: &Self::Transaction,
+		block: Self::BlockHash,
+		proof: &Self::TransactionInclusionProof,
+	) -> bool;
+}
+
+/// The module configuration trait
+pub trait Trait: frame_system::Trait {
+	/// Peer blockchain type.
+	type PeerBlockchain: Blockchain;
+	/// Peer blockchain transaction parser.
+	type PeerMaybeLockFundsTransaction: MaybeLockFundsTransaction<
+		Transaction = <Self::PeerBlockchain as Blockchain>::Transaction,
+	>;
+	/// Map between blockchains recipients.
+	type RecipientsMap: RecipientsMap<
+		PeerRecipient = <Self::PeerMaybeLockFundsTransaction as MaybeLockFundsTransaction>::Recipient,
+		Recipient = Self::AccountId,
+	>;
+	/// This blockchain currency amount type.
+	type Amount;
+	/// Converter from peer blockchain currency type into current blockchain currency type.
+	type CurrencyConverter: CurrencyConverter<
+		SourceAmount = <Self::PeerMaybeLockFundsTransaction as MaybeLockFundsTransaction>::Amount,
+		TargetAmount = Self::Amount,
+	>;
+	/// Something that could grant money.
+	type Airdrop: Airdrop<Recipient = Self::AccountId, Amount = Self::Amount>;
```

Comment:
We should just require `Currency` trait for both this and `Amount`. I think it has a nice (and commonly used within substrate) API to mint new tokens and is implemented for `balances` module already.

Reply:
Nevermind this - probably the way we create tokens will differ depending on the chain. I'd suggest naming `Airdrop` something more generic like: `type TransferTokens` so that it doesn't imply the tokens are "created" out of thin air, but rather they might be transferred from some common pool.

Reply:
Renamed to `trait DepositInto`